backend/manager/modules/bll/src/main/java/org/ovirt/engine/core/bll/exportimport/ConvertVmCommand.java

```diff
@@ -132,9 +137,44 @@ private ConvertVmVDSParameters buildConvertParameters() {
         return parameters;
     }
 
+    private StorageDomainType getIsoStorageDomainType() {
+        if (isoStorageDomainType == null) {
+            isoStorageDomainType = getParameters().getVirtioIsoStorageDomainId() != null
+                    ? storageDomainDao.get(getParameters().getVirtioIsoStorageDomainId()).getStorageDomainType()
+                    : StorageDomainType.ISO;
+        }
+        return isoStorageDomainType;
+    }
+
     protected String getVirtioIsoPath() {
-        return getParameters().getVirtioIsoName() == null ? null :
-            new File(getIsoPrefix(getStoragePoolId(), getVdsId()), getParameters().getVirtioIsoName()).getPath();
+        if (getParameters().getVirtioIsoName() == null) {
+            return null;
+        }
+        if (getIsoStorageDomainType() == StorageDomainType.ISO) {
+            return new File(getIsoPrefix(getStoragePoolId(), getVdsId()), getParameters().getVirtioIsoName()).getPath();
+        }
+        DiskImage isoImage = imagesHandler.getSnapshotLeaf(new Guid(getParameters().getVirtioIsoName()));
+        PrepareImageReturn preparedImage =
+                (PrepareImageReturn) imagesHandler
+                        .prepareImage(getStoragePoolId(),
+                                getParameters().getVirtioIsoStorageDomainId(),
+                                isoImage.getId(),
+                                isoImage.getImageId(),
+                                getVdsId())
+                        .getReturnValue();
+        return preparedImage.getImagePath();
+    }
+
+    private void teardownVirtioIso() {
+        if (getParameters().getVirtioIsoName() == null || getIsoStorageDomainType() == StorageDomainType.ISO) {
+            return;
+        }
+        DiskImage isoImage = imagesHandler.getSnapshotLeaf(new Guid(getParameters().getVirtioIsoName()));
+        imagesHandler.teardownImage(getStoragePoolId(),
```

Comment:
and what if this iso is used concurrently by another VM on this host?

Reply:
Right. So, I can block this, like it is done in `MeasureVolumeCommand` - it fails in validation, if the disk is used by a running VM.

But there is also a case when the same ISO is used by several `ConvertVm` commands. I can use in-memory locks for this, for disks on data domains only.

Reply:
So it means we cannot attach the ISO if is used by an active VM?  Does anything prevents using it by multiple active VMs not considering the import?  Is it a limitation of prepare / tear down image on the Vdsm side, perhaps missing support for using and sharing read-only images (how do we use CDs if it is the case? or can we not use CDs from data domains)?  Please explain in the commit message what we reasonably can do, cannot do and actually do here.

Reply:
Done.

Reply:
Another question is: what will happen, if some VM using the ISO image will be started **after** the `ConvertVmCommand` is started? I don't see this solved for `MeasureVolumeCommand` also. And cannot figure out a good way to prevent this.

Reply:
@mz-pdm In `MeasureVolumeCommand` there was a different reason for blocking usage of the same disk by any active VMs. Not because of `prepare/teardownImage()`, but because qemu cannot use a disk that was open for writing by another process. If a VM using the disk starts after starting `MeasureVmCommand`, qemu will fail to start.

In our case the ISO disk will be opened for reading, but the limitation of `prepare/teardownImage()` still exists. Although, it is quite unlikely that some VM will use the ISO with virtio drivers as a disk.

Reply:
There may be other ISOs that users might use in running VMs.  If I understand you correctly, there is nothing preventing attempts to start multiple VMs using the same ISO, but it will fail in Vdsm on preparing the image.

If we cannot resolve it easily, we could perhaps do:

- Providing a doc text explaining the limitations.

- Does or could the user get a comprehensible error message on such a VM start failure?

Reply:
@mz-pdm I've talked with Benny about `prepare/teardownImage()`. There are two important things:

1. `teardownImage()` does not affect running VMs using the same disk. This situation is handled inside VDSM since 4.4.
2. It is true that several commands cannot use `prepare/teardownImage()` on the same disk simultaneously. But this is true for block storage only. It's not a problem for file storage.

I'll update the patch accordingly.

Reply:
Great, good news, thanks.